components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1664,6 +1802,47 @@ where
                 self.finish_unsafe_recovery_wait_apply();
             }
         }
+        if self.fsm.unsafe_recovery_wait_failed_voters_demotion_apply {
+            if self.fsm.unsafe_recovery_num_tasks.is_none() {
+                panic!(
+                    "{:?} is waiting for failed peers' demotion to apply, but the number of total
+		    recovery tasks is not set",
+                    self.fsm.peer.region()
+                );
+            }
+            self.fsm.unsafe_recovery_wait_failed_voters_demotion_apply = false;
+            let num_tasks = self.fsm.unsafe_recovery_num_tasks.take().unwrap();
+
+            // Exit joint state and wait for it to be applied.
+            let region = self.fsm.peer.region();
+            let region_id = region.get_id();
+            let mut req = new_admin_request(region_id, self.fsm.peer.peer.clone());
+            req.mut_header()
+                .set_region_epoch(region.get_region_epoch().clone());
+            let mut admin = AdminRequest::default();
+            admin.set_cmd_type(AdminCmdType::ChangePeerV2);
+            let cpv2 = ChangePeerV2Request::default();
+            admin.set_change_peer_v2(cpv2);
+            req.set_admin_request(admin);
+            self.propose_raft_command_internal(
```

Comment:
use `ctx.router.send_raft_command`?

Reply:
We need to know the target commit index immediately, so probably we should keep it this way.